purchases/src/androidTestIntegrationTest/kotlin/com/revenuecat/purchases/OfflineEntitlementsIntegrationTest.kt

```diff
@@ -0,0 +1,211 @@
+                    },
+                    onSuccess = {
+                        latch.countDown()
+                        fail("Expected error but got success: $it")
+                    }
+                )
+            }
+        }
+    }
+
+    @Test
+    fun entersOfflineEntitlementsModeIfPurchaseRequestReturns500() {
+        val storeProduct = StoreProductFactory.createGoogleStoreProduct(
+            productId = testProductId,
+            basePlanId = testBasePlanId,
+            subscriptionOptionsList = listOf(
+                StoreProductFactory.createGoogleSubscriptionOption(
+                    productId = testProductId,
+                    basePlanId = testBasePlanId,
+                )
+            )
+        )
+
+        ensureBlockFinishes { latch ->
+            setupTest(
+                buildSharedPreferencesMap(shouldIncludeCustomerInfo = true),
+                initialActivePurchases = emptyMap()
+            ) { activity ->
+                val receivedCustomerInfos = mutableListOf<CustomerInfo>()
+                Purchases.sharedInstance.updatedCustomerInfoListener = UpdatedCustomerInfoListener {
+                    receivedCustomerInfos.add(it)
+                }
+                every {
+                    mockBillingAbstract.makePurchaseAsync(any(), any(), any(), any(), any(), any())
+                } answers {
+                    mockActivePurchases(initialActivePurchases)
+                    latestPurchasesUpdatedListener!!.onPurchasesUpdated(listOf(initialActiveTransaction))
+                }
+                mockBillingAbstract.mockQueryProductDetails(queryProductDetailsSubsReturn = listOf(storeProduct))
+                Purchases.sharedInstance.purchaseWith(
+                    PurchaseParams.Builder(activity, storeProduct).build(),
+                    onError = { error, _ ->
+                        latch.countDown()
+                        fail("Expected success but got error: $error")
+                    },
+                    onSuccess = { _, customerInfo ->
+                        assertThat(customerInfo.entitlements.active.keys).containsExactly(
+                            "pro_cat"
+                        )
+                        assertThat(receivedCustomerInfos).hasSize(2)
+                        assertThat(receivedCustomerInfos.first().entitlements.active).isEmpty()
+                        assertThat(receivedCustomerInfos.last().entitlements.active.keys).containsExactly("pro_cat")
+                        latch.countDown()
+                    }
+                )
+            }
+        }
+    }
+
+    // region helpers
+
+    private fun setupMockServer() {
+        server = MockWebServer()
+        server.start()
+        Purchases.proxyURL = server.url("").toUrl()
+        server.dispatcher = object : Dispatcher() {
+            override fun dispatch(request: RecordedRequest): MockResponse {
+                return MockResponse().setResponseCode(RCHTTPStatusCodes.ERROR)
```

Comment:
Right now this is always returning 500. The plan would be to change this to also return successes in some endpoints for some tests to test recovery.